From 24771a192ece991be650064a87765dd926553272 Mon Sep 17 00:00:00 2001
From: Markus Niebel <Markus.Niebel@tq-group.com>
Date: Fri, 10 May 2019 10:27:49 +0200
Subject: [PATCH 1/1] asoc: imx-tlv320aic32x4: add widgets and routing

- Widgets common for all MBa<6x,6ulx,6ullx,7x,8Xx,8Mx> starterkits
  * Headphone
  * Line In / Out
  * Mic in
- Routing common for all MBa<6x,6ulx,6ullx,7x,8Xx,8Mx> starterkits
  * Headphone -> HPL/R
  * Line In -> LIN_1_L/R
  * Mic in -> LIN_3
  * Line Out -> LOL/R

Todo: should better go to DT

Signed-off-by: Markus Niebel <Markus.Niebel@tq-group.com>
---
 sound/soc/fsl/imx-tlv320aic32x4.c | 26 +++++++++++++++++++++++++-
 1 file changed, 25 insertions(+), 1 deletion(-)

diff --git a/sound/soc/fsl/imx-tlv320aic32x4.c b/sound/soc/fsl/imx-tlv320aic32x4.c
index 25a1365ec81d..8348474e5d3a 100644
--- a/sound/soc/fsl/imx-tlv320aic32x4.c
+++ b/sound/soc/fsl/imx-tlv320aic32x4.c
@@ -29,6 +29,26 @@ struct imx_tlv320aic_data {
 	unsigned int clk_frequency;
 };
 
+static const struct snd_soc_dapm_widget imx_tlv320aic_dapm_widgets[] = {
+	SND_SOC_DAPM_MIC("Mic Jack", NULL),
+	SND_SOC_DAPM_LINE("Line In Jack", NULL),
+	SND_SOC_DAPM_HP("Headphone Jack", NULL),
+	SND_SOC_DAPM_SPK("Line Out Jack", NULL),
+};
+
+static const struct snd_soc_dapm_route imx_tlv320aic_dapm_routes[] = {
+	{"Mic Bias", NULL, "Mic Jack"},
+	{"IN1_R", NULL, "Line In Jack"},
+	{"IN1_L", NULL, "Line In Jack"},
+	{"IN3_R", NULL, "Mic Bias"},
+	{"IN3_L", NULL, "Mic Bias"},
+
+	{"Headphone Jack", NULL, "HPL"},
+	{"Headphone Jack", NULL, "HPR"},
+	{"Line In Jack", NULL, "LOL"},
+	{"Line In Jack", NULL, "LOR"},
+};
+
 static int imx_tlv320aic_dai_init(struct snd_soc_pcm_runtime *rtd)
 {
 	struct imx_tlv320aic_data *data = container_of(rtd->card,
@@ -103,8 +123,12 @@ static int imx_tlv320aic_probe(struct platform_device *pdev)
 	data->dai.init = &imx_tlv320aic_dai_init;
 	data->dai.dai_fmt = SND_SOC_DAIFMT_I2S | SND_SOC_DAIFMT_NB_NF |
 			    SND_SOC_DAIFMT_CBM_CFM;
-
+	data->card.dapm_widgets = imx_tlv320aic_dapm_widgets;
+	data->card.num_dapm_widgets = ARRAY_SIZE(imx_tlv320aic_dapm_widgets);
+	data->card.dapm_routes	= imx_tlv320aic_dapm_routes,
+	data->card.num_dapm_routes = ARRAY_SIZE(imx_tlv320aic_dapm_routes),
 	data->card.dev = &pdev->dev;
+
 	ret = snd_soc_of_parse_card_name(&data->card, "model");
 	if (ret)
 		goto fail;
